ui-components/src/page_components/listing/ListingCard.tsx

```diff
@@ -22,10 +26,80 @@ export interface ListingCardProps {
   tableHeaderProps?: ListingCardHeaderProps
   tableProps: ListingCardTableProps
   detailsLinkClass?: string
+  listingID: string
 }
 
 const ListingCard = (props: ListingCardProps) => {
   const { imageCardProps, tableProps, detailsLinkClass, tableHeaderProps } = props
+  const { profile, userProfileService } = useContext(AuthContext)
+  const preferences: UserPreferences = profile?.preferences || {
+    sendEmailNotifications: false,
+    sendSmsNotifications: false,
+    favoriteIDs: [],
+  }
+  const [favoriteState, updateFavoriteState] = useState("favored")
+
+  useEffect(() => {
+    if (preferences.favoriteIDs?.includes(props.listingID)) {
+      updateFavoriteState("favored")
+    } else {
+      updateFavoriteState("notFavored")
+    }
+  }, [preferences.favoriteIDs, props.listingID])
+
+  const addToFavorite = async () => {
+    if (!profile) {
+      return
+    }
+    updateFavoriteState("favored")
+    const preferences: UserPreferences = profile.preferences || {
+      sendEmailNotifications: false,
+      sendSmsNotifications: false,
+      favoriteIDs: [],
+    }
+
+    if (preferences.favoriteIDs?.includes(props.listingID)) {
+      return
+    } else {
+      preferences.favoriteIDs?.push(props.listingID)
+      console.log("Adding Listing")
+    }
+
+    try {
+      await userProfileService?.update({
+        body: { ...profile, preferences },
+      })
+    } catch (err) {
+      console.warn(err)
+    }
+  }
+
+  const removeFavorite = async () => {
+    if (!profile) {
+      return
+    }
+    updateFavoriteState("notFavored")
+    const preferences: UserPreferences = profile.preferences || {
+      sendEmailNotifications: false,
+      sendSmsNotifications: false,
+      favoriteIDs: [],
+    }
+    if (!preferences.favoriteIDs || preferences?.favoriteIDs?.length === 0) {
+      return
+    }
+
+    const index: number = preferences.favoriteIDs?.indexOf(props.listingID) || 0
```

Comment:
I don't think we should fallback to 0 here or else in the case where we don't have a listingId we'll end up removing the first item in the list. Instead we should early return (and maybe a log an error) since we can't proceed if the favoriteIDs already doesn't have the id

Reply:
updated